src/library/scala/collection/mutable/RedBlackTree.scala

```diff
@@ -0,0 +1,545 @@
+   * therefore, the last node), this method returns `null`.
+   */
+  private[this] def successor[A, B](node: Node[A, B]): Node[A, B] = {
+    if (node.right ne null) minNodeNonNull(node.right)
+    else {
+      var x = node
+      var y = x.parent
+      while ((y ne null) && (x eq y.right)) {
+        x = y
+        y = y.parent
+      }
+      y
+    }
+  }
+
+  /**
+   * Returns the node that precedes `node` in an in-order tree traversal. If `node` has the minimum key (and is,
+   * therefore, the first node), this method returns `null`.
+   */
+  private[this] def predecessor[A, B](node: Node[A, B]): Node[A, B] = {
+    if (node.left ne null) maxNodeNonNull(node.left)
+    else {
+      var x = node
+      var y = x.parent
+      while ((y ne null) && (x eq y.left)) {
+        x = y
+        y = y.parent
+      }
+      y
+    }
+  }
+
+  private[this] def rotateLeft[A, B](tree: Tree[A, B], x: Node[A, B]): Unit = if (x ne null) {
+    // assert(x.right ne null)
+    val y = x.right
+    x.right = y.left
+
+    if (y.left ne null) y.left.parent = x
+    y.parent = x.parent
+
+    if (x.parent eq null) tree.root = y
+    else if (x eq x.parent.left) x.parent.left = y
+    else x.parent.right = y
+
+    y.left = x
+    x.parent = y
+  }
+
+  private[this] def rotateRight[A, B](tree: Tree[A, B], x: Node[A, B]): Unit = if (x ne null) {
+    // assert(x.left ne null)
+    val y = x.left
+    x.left = y.right
+
+    if (y.right ne null) y.right.parent = x
+    y.parent = x.parent
+
+    if (x.parent eq null) tree.root = y
+    else if (x eq x.parent.right) x.parent.right = y
+    else x.parent.left = y
+
+    y.right = x
+    x.parent = y
+  }
+
+  /**
+   * Transplant the node `from` to the place of node `to`. This is done by setting `from` as a child of `to`'s previous
+   * parent and setting `from`'s parent to the `to`'s previous parent. The children of `from` are left unchanged.
+   */
+  private[this] def transplant[A, B](tree: Tree[A, B], to: Node[A, B], from: Node[A, B]): Unit = {
+    if (to.parent eq null) tree.root = from
+    else if (to eq to.parent.left) to.parent.left = from
+    else to.parent.right = from
+
+    if (from ne null) from.parent = to.parent
+  }
+
+  // ---- tree traversal ----
+
+  def foreach[A, B, U](tree: Tree[A, B], f: ((A, B)) => U): Unit = foreachNode(tree.root, f)
+
+  private[this] def foreachNode[A, B, U](node: Node[A, B], f: ((A, B)) => U): Unit =
+    if (node ne null) foreachNodeNonNull(node, f)
+
+  private[this] def foreachNodeNonNull[A, B, U](node: Node[A, B], f: ((A, B)) => U): Unit = {
+    if (node.left ne null) foreachNodeNonNull(node.left, f)
+    f((node.key, node.value))
+    if (node.right ne null) foreachNodeNonNull(node.right, f)
+  }
+
+  def transform[A, B](tree: Tree[A, B], f: (A, B) => B): Unit = transformNode(tree.root, f)
+
+  private[this] def transformNode[A, B, U](node: Node[A, B], f: (A, B) => B): Unit =
+    if (node ne null) transformNodeNonNull(node.left, f)
```

Comment:
Should this be `transformNodeNonNull(node, f)`?  The way it is gives a NPE and misses part of the tree.